File: 1/JAVA/Ackermann_a.java
public class Ackermann {

    public static int ackermannA(int m, int n) {
        if (m == 0) return n + 1;
        if (m > 0 && n == 0) return ackermannA(m - 1, 1);
        return ackermannA(m - 1, ackermannA(m, n - 1));
    }

    public static void main(String[] args) {
        int m = 2, n = 3;
        System.out.println("A(" + m + ", " + n + ") = " + ackermannA(m, n));
    }
}

File: 2/PROLOG/ackermann_a.pl
:- use_module(library(statistics)).

ackermann_a(0, N, Result) :-
    Result is N + 1.
ackermann_a(M, 0, Result) :-
    M > 0,
    M1 is M - 1,
    ackermann_a(M1, 1, Result).
ackermann_a(M, N, Result) :-
    M > 0,
    N > 0,
    M1 is M - 1,
    N1 is N - 1,
    ackermann_a(M, N1, TempResult),
    ackermann_a(M1, TempResult, Result).

compute_a :-
    between(0, 3, M),
    between(0, 3, N),
    time((ackermann_a(M, N, Result), write('A('), write(M), write(', '), write(N), write(') = '), write(Result), nl)),
    fail.
compute_a.
